codegen/ace/src/circuit.rs

```diff
@@ -0,0 +1,507 @@
+    }
+
+    /// Append an operation to the circuit after applying some optimizations
+    /// from the algebraic laws of addition, multiplication and subtraction.
+    pub fn append(&mut self, op: Operation) -> NodeIndex {
+        use Operation::*;
+        match op {
+            Value(_) => self.append_cached(op),
+            Add(l, r) => {
+                if l == self.zero {
+                    r
+                } else if r == self.zero {
+                    l
+                } else {
+                    // indexes are sorted so that semantically equivalent nodes
+                    // (because of commutativity) are syntactically equivalent
+                    // and can be cached
+                    let (l, r) = if l < r { (l, r) } else { (r, l) };
+                    self.append_cached(Add(l, r))
+                }
+            }
+            Mul(l, r) => {
+                if l == self.one {
+                    r
+                } else if r == self.one {
+                    l
+                } else if l == self.zero || r == self.zero {
+                    self.zero
+                } else {
+                    // indexes are sorted so that semantically equivalent nodes
+                    // (because of commutativity) are syntactically equivalent
+                    // and can be cached
+                    let (l, r) = if l < r { (l, r) } else { (r, l) };
+                    self.append_cached(Mul(l, r))
+                }
+            }
+            Sub(l, r) => {
+                if r == self.zero {
+                    l
+                } else if l == r {
+                    self.zero
+                } else {
+                    self.append_cached(Sub(l, r))
+                }
+            }
+        }
+    }
+
+    // Given that the TraceAccesses of the Air are on maximum two
+    // segments (main and aux), the inputs needed by the ACE
+    // chiplets are added on a 3rd segment so that they are
+    // ordered last. Anyway the segment information is lost when
+    // the circuit is serialized.
+    pub fn input(&mut self) -> NodeIndex {
+        let res = self.append(Operation::Value(Value::TraceAccess(TraceAccess::new(
+            3,
+            0,
+            self.n_inputs,
+        ))));
+        self.n_inputs += 1;
+        res
+    }
+
+    pub fn constant(&mut self, c: u64) -> NodeIndex {
+        self.append(Operation::Value(Value::Constant(c)))
+    }
+
+    pub fn add(&mut self, l: NodeIndex, r: NodeIndex) -> NodeIndex {
+        self.append(Operation::Add(l, r))
+    }
+
+    pub fn mul(&mut self, l: NodeIndex, r: NodeIndex) -> NodeIndex {
+        self.append(Operation::Mul(l, r))
+    }
+
+    pub fn sub(&mut self, l: NodeIndex, r: NodeIndex) -> NodeIndex {
+        self.append(Operation::Sub(l, r))
+    }
+
+    /// `\sum_i vec[i]`
+    pub fn sum(&mut self, vec: &[NodeIndex]) -> NodeIndex {
+        assert!(!vec.is_empty());
+        vec.iter().fold(self.zero, |acc, r| self.add(acc, *r))
```

Comment:
I think adding zero is optimized, but `reduce` ensures that we don't perform a trivial addition.

```suggestion
        vec.iter().reduce(|acc, r| self.add(acc, *r)).unwrap_op(self.zero)
```

Reply:
Nice to learn about reduce. I tried but it gets messy with the references. I would leave it a fold for now.